app/Platform/Actions/ResolveAchievementSetsAction.php

```diff
@@ -0,0 +1,201 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Platform\Actions;
+
+use App\Models\GameAchievementSet;
+use App\Models\GameHash;
+use App\Models\User;
+use App\Models\UserGameAchievementSetPreference;
+use App\Platform\Enums\AchievementSetType;
+use Illuminate\Support\Collection;
+
+class ResolveAchievementSetsAction
+{
+    /**
+     * Resolves the achievement sets for a given game hash and user.
+     *
+     * @return Collection<int, GameAchievementSet>
+     */
+    public function execute(GameHash $gameHash, User $user): Collection
+    {
+        $game = $gameHash->game;
+
+        $initialSet = GameAchievementSet::where('game_id', $game->id)->first();
+        if (!$initialSet) {
+            return collect();
+        }
+
+        // Determine the current game context based on the initial set and game.
+        [$gameIdToUse, $typesToLoad] = $this->buildResolutionContext($initialSet, $game->id);
+
+        // Retrieve all relevant achievement sets for the current context.
+        $allSets = $this->getAllAchievementSets($gameIdToUse, $typesToLoad);
+
+        // Fetch user opt-in/out preferences for these sets.
+        $userSetPreferences = $this->getUserSetPreferences($user, $allSets);
+
+        // Filter the sets based on preferences and hash compatibility.
+        $filteredSets = $this->filterSets($allSets, $gameHash, $user, $userSetPreferences);
+
+        // Sort the sets based on their type priority.
+        $sortedSets = $this->sortSets($filteredSets);
+
+        // Attach core_game_id to each set, which may be used later to resolve
+        // things such as leaderboard definitions.
+        $this->attachCoreGameIds($sortedSets);
+
+        return $sortedSets->values();
+    }
+
+    /**
+     * Given a GameAchievementSet, returns the set's game ID and the types to load.
+     *
+     * @return array [int $gameIdToUse, array|null $typesToLoad]
+     */
+    private function buildResolutionContext(GameAchievementSet $initialSet, int $defaultGameId): array
+    {
+        $links = GameAchievementSet::where('achievement_set_id', $initialSet->achievement_set_id)->get();
+
+        $exclusiveLink = $links->firstWhere('type', AchievementSetType::Exclusive);
+        $specialtyLink = $links->firstWhere('type', AchievementSetType::Specialty);
+        $bonusLink = $links->firstWhere('type', AchievementSetType::Bonus);
```

Comment:
Suggest moving these down next to their `if`s. 

I realize the query only runs once and these are filters, but each of the `if`s can short-circuit out, and I prefer to keep the logic next to the usage.

Reply:
Updated in latest.